File: Divide And Conquer/points_and_segments.py
import sys

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    m = data[1]
    starts = data[2:2 * n + 2:2]
    ends = data[3:2 * n + 2:2]
    points = data[2 * n + 2:]
    all = list(map(lambda x: (x, 1), starts))
    all += list(map(lambda x: (x, 3), ends))
    all += list(map(lambda x: (x, 2), points))

    all.sort()
    count = 0
    record = {}
    for point in all:
        if point[1] == 1:
            count += 1
        elif point[1] == 3:
            count -= 1
        else:
            record[point[0]] = count

    for x in points:
        print(record[x], end=' ')

File: Divide And Conquer/points_and_segments.cpp
/**
 * Given a set of points on a line and a set of segments on a line. The goal is
   to compute, for each point, the number of segments that contain this point.
 * @file points_and_segments.cpp
 * @author Bhed Kumar Kushwaha
 * @version 1.0
 */

#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <utility>
#include <vector>

using std::pair;
using std::unordered_map;
using std::vector;

int main() {
    int n, m;
    std::cin >> n >> m;
    vector<pair<int, char>> all(n + n + m);

    for (size_t i = 0; i < n; i++) {
        std::cin >> all[i].first >> all[n + i].first;
        all[i].second     = 'l';
        all[n + i].second = 'r';
    }
    vector<int> points(m);
    for (size_t i = 0; i < m; i++) {
        std::cin >> points[i];
        all[n + n + i].first  = points[i];
        all[n + n + i].second = 'p';
    }

    std::sort(all.begin(), all.end(), [](const auto &i, const auto &j) {
        if (i.first == j.first) {
            if (i.second == 'p' && j.second == 'l')
                return false;
            else if (i.second == 'r' && j.second == 'p')
                return false;
            else if (i.second == 'r' && j.second == 'l')
                return false;
            return true;
        }
        return i.first < j.first;
    });

    int count{};
    unordered_map<int, int> result;
    for (auto e : all) {
        if (e.second == 'l') {
            count++;
        } else if (e.second == 'p') {
            result[e.first] = count;
        } else {
            count--;
        }
    }

    for (size_t i = 0; i < m; i++) {
        std::cout << result[points[i]] << ' ';
    }
}
